src/torchjd/autojac/mtl_backward.py

```diff
@@ -23,12 +25,92 @@
 )
 
 
+def _get_tensors_memory_id(tensor_list: list[Tensor]) -> set[int]:
+    """
+    Based on a list of lists of tensor objects returns the unique memory adress of all
+    :param tensor_list: list containing pytorch Tensor objects
+    """
+    return set([id(tensor) for tensor in tensor_list])
+
+
+def _traverse_ag_graph(grad_graph: Node) -> list[Tensor]:
+    """
+    Traverses an autograd graph iteratively and extracts all variables (trainable parameters)
+    that lead to the given node.
+
+    :param grad_graph: An autograd computation node.
+    :return: A list of tensors representing the variables that lead to the given node.
+    """
+    stack = [grad_graph]
+    output_params = []
+
+    while stack:
+        current_node = stack.pop()
+
+        child_nodes = current_node.next_functions
+        for child in child_nodes:
+            if child[0] is None:
+                # Skip constants in the calculation
+                continue
+            if hasattr(child[0], "variable"):
+                output_params.append(child[0].variable)
+            else:
+                stack.append(child[0])
+
+    return output_params
+
+
+def _determine_shared_not_shared(
+    tensor_list: list[list[Tensor]],
```

Comment:
I think we may want `tensors_list`, or maybe there is a better name for that?

Reply:
tensor_lists or tensors_lists (there are multiple lists)

Reply:
You are right, also we tend to not have `list` or `set` in names of variables, maybe `nested_tensors`?

Reply:
I think that `list[set[Tensor]]` would be appropriate, it is a list whose indices correspond to the losses, and elements are unordered and disjoint collections of tensors that appear in the graph of the loss.

Reply:
Let's call it `tensor_sets` then. This is more consistent with the rest of torchjd.